doc/langref.html.in

```diff
@@ -6899,8 +6899,255 @@ fn foo() i32 {
 
       {#header_open|Result Location Semantics#}
       <p>
-      <a href="https://github.com/ziglang/zig/issues/2809">TODO add documentation for this</a>
+      During compilation, every Zig expression and sub-expression is assigned optional result location
+      information. This information dictates what type the expression should have (its result type), and
+      where the resulting value should be placed in memory (its result location). This is not an
+      implementation detail: this logic is codified into the Zig language specification. This system is
+      collectively referred to as "Result Location Semantics".
       </p>
+      {#header_open|Result Types#}
+      <p>
+      Result types are propagated recursively through expressions where possible. For instance, if the
+      expression {#syntax#}&e{#endsyntax#} has result type {#syntax#}*u32{#endsyntax#}, then
+      {#syntax#}e{#endsyntax#} is given a result type of {#syntax#}u32{#endsyntax#}, allowing the
+      language to perform this coercion before taking a reference.
+      </p>
+      <p>
+      The result type mechanism is utilized by casting builtins such as {#syntax#}@intCast{#endsyntax#}.
+      Rather than taking as an argument the type to cast to, these builtins use their result type to
+      determine this information. The result type is often known from context; where it is not, the
+      {#syntax#}@as{#endsyntax#} builtin can be used to explicitly provide a result type.
+      </p>
+      <p>
+      We can break down the result types for each component of a simple expression as follows:
+      </p>
+      {#code_begin|test|result_type_propagation#}
+test "result type propagates through struct initializer" {
+    const S = struct { x: u32 };
+    const val: u64 = 123;
+    const s: S = .{ .x = @intCast(val) };
+    // .{ .x = @intCast(val) }   has result type `S` due to the type annotation
+    //         @intCast(val)     has result type `u32` due to the type of the field `S.x`
+    //                  val      has no result type, as it is permitted to be any integer type
+    try expectEqual(@as(u32, 123), s.x);
+}
+      {#code_end#}
+      <p>
+      This result type information is useful for the aforementioned cast builtins, as well as to avoid
+      the construction of pre-coercion values, and to avoid the need for explicit type coercions in some
+      cases. The following table details how some common expressions propagate result types, where
+      {#syntax#}x{#endsyntax#} and {#syntax#}y{#endsyntax#} are arbitrary sub-expressions.
```

Comment:
This sentence reads like it is accidentally incomplete. Is there a complete table anywhere, ie as issue in the language specification repo? Would the algorithm be more intuitive?

Reply:
I'm not sure what seems incomplete here, but I might be missing something - could you elaborate? I intentionally only included a small set of examples, as detailing this for the whole Zig syntax would be fairly broad. If you can think of other important examples which should be in the table, let me know.

In terms of referencing the (currently non-existent) language specification, RLS is a little more complicated than the docs let on. The concepts of result types and result locations are somewhat interleaved; the address-of operation is a little subtle since Zig doesn't have the concept of an unknown-length array type; type-inferred pointers work differently; and destructuring syntax has direct ties to RLS. We could definitely expand on some of this in the langref, but I don't think it's a good idea to bombard the user with all the technical details immediately.

Reply:
> but I might be missing something - could you elaborate?

Usually the phrase "some" is elsewhere clarified in what is meant, see for example [here](https://ziglang.org/documentation/master/#toc-Explicit-Casts). It would be nice to reference the language specification or an issue for the formal specification / necessary context like how it is done elsewhere in the language reference.